src/lightning/pytorch/strategies/xla_fsdp.py

```diff
@@ -0,0 +1,587 @@
+        optimizer shards stored per file. If the user specifies full checkpointing, the directory will contain a
+        consolidated checkpoint combining all of the sharded checkpoints.
+
+        """
+        if not _TORCH_GREATER_EQUAL_2_0:
+            raise NotImplementedError(
+                "Saving and loading checkpoints with the `XLAFSDPStrategy` is not supported in PyTorch < 2.0."
+                " Please upgrade `torch`."
+            )
+        # broadcast the path from rank 0 to ensure all the states are saved in a common path
+        path = Path(self.broadcast(filepath))
+        if path.is_dir() and any(path.iterdir()):
+            raise FileExistsError(f"The checkpoint directory already exists and is not empty: {path}")
+
+        converted_state = {"model": checkpoint.pop("state_dict")}
+        converted_state.update({"optimizer": checkpoint.pop("optimizer_states")})
+        import torch_xla.core.xla_model as xm
+
+        # ensure model parameters are updated
+        xm.mark_step()
+
+        parallel_devices = self.parallel_devices
+        assert parallel_devices is not None
+        if self._sequential_save:
+            # each host runs this in parallel, but the ranks in the host run it sequentially
+            for rank in range(len(parallel_devices)):
+                if rank == self.local_rank:
+                    self._save_checkpoint_shard(path, converted_state, storage_options)
+                self.barrier(f"wait-for-{rank}-save")
+        else:
+            self._save_checkpoint_shard(path, converted_state, storage_options)
+
+        if self._state_dict_type == "full":
+            ckpt_prefix = str(path / "checkpoint")
+            ckpt_suffix = "_rank-*-of-*.pth"
+            if len(parallel_devices) != self.world_size:  # multihost
+                raise OSError(
+                    "Multihost setups do not have a shared filesystem, so the checkpoint shards cannot be consolidated"
+                    " into a single checkpoint after saving them. Please switch to"
+                    " `XLAFSDPStrategy(state_dict_type='sharded')`. TIP: You can consolidate them manually by getting"
+                    " them together into a single directory and running `python -m"
+                    f" torch_xla.distributed.fsdp.consolidate_sharded_ckpts --ckpt_prefix {ckpt_prefix!r} --ckpt_suffix"
+                    f" {ckpt_suffix!r} --save_path 'path/to/consolidated.ckpt'`."
+                )
+
+            from torch_xla.distributed.fsdp import consolidate_sharded_model_checkpoints
+
+            self.barrier("before_ckpt_consolidation")
+            if self.is_global_zero:
+                save_path = path.parent / "consolidated.ckpt"
+                # save consolidated checkpoint separate to the shards
+                consolidate_sharded_model_checkpoints(ckpt_prefix, ckpt_suffix, str(save_path))
```

Comment:
Currently,`XLAFSDP` does not support full model checkpointing (via `consolidate_sharded_model_checkpoints`) since `shard_metadata` isn't saved when saving the individual shards. Lines 348 and 349, we only save the model and the optimizer, but not the `shard_metadata`, as seen in (https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist_fsdp_with_ckpt.py#L289)

From my digging, it appears if we also want to save `shard_metadata` we would need to edit `dump_sharding` in https://github.com/Lightning-AI/lightning/blob/0777aab31e07748b9f1f5bda73f6a6fa936cb54d/src/lightning/pytorch/trainer/connectors/checkpoint_connector.py#L402. However, I'm not sure what impact this will have on the other strategies and their checkpoint saving.

@carmocca what are your thoughts here?

Reply:
In the interest of time, I suggest that we don't support this now and raise a `NotImplementedError` accordingly.

We know that the current trainer design where the strategy cannot have a part in creating the state dict is flawed for the sharded strategies. It will need a more holistic solution than adding a special condition just for this strategy.

Reply:
Sounds good, will make some adjustments

Reply:
added function `XLAFSDPStrategy.lightning_module_state_dict()` that includes model state dict, shard metadata, and optimizer state dict.

https://github.com/Lightning-AI/lightning/blob/d7d78f16a1a1c08dfeea45db7c42a06cfb4fb368/src/lightning/pytorch/strategies/xla_fsdp.py#L468-L475